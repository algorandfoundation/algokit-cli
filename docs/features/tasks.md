# AlgoKit Tasks

AlgoKit Tasks are a collection of handy tasks that can be used to perform various operations on Algorand blockchain.

## Features

- [Wallet Aliasing](./tasks/wallet.md) - Manage your Algorand addresses and accounts effortlessly with the AlgoKit Wallet feature. This feature allows you to create short aliases for your addresses and accounts on AlgoKit CLI.
- [Vanity Address Generation](./tasks/vanity_address.md) - Generate vanity addresses for your Algorand accounts with the AlgoKit Vanity feature. This feature allows you to generate Algorand addresses which contains a specific keyword of your choice.
- [Transfer Assets or Algos](./tasks/transfer.md) - Transfer Algos or Assets from one account to another with the AlgoKit Transfer feature. This feature allows you to transfer Algos or Assets from one account to another on Algorand blockchain.
- [Opt-(in|out) Assets](./tasks/opt.md) - Opt-in or opt-out of Algorand Asset(s). Supports single or multiple assets.
- [Signing transactions](./tasks/sign.md) - Sign goal clerk compatible Algorand transactions.
- [Sending transactions](./tasks/send.md) - Send signed goal clerk compatible Algorand transactions.
- [NFD lookups](./tasks/nfd.md) - Perform a lookup via NFD domain or address, returning the associated address or domain respectively using the AlgoKit CLI.
- [IPFS uploads](./tasks/ipfs.md) - Upload files to IPFS.
- [Asset minting](./tasks/mint.md) - Mint new fungible or non-fungible assets on Algorand.
- [Analyze TEAL code](./tasks/analyze.md) - Analyze TEAL code using [`tealer`](https://github.com/crytic/tealer) integration for common vulnerabilities.
