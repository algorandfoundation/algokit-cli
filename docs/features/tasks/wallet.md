# AlgoKit Task Wallet

Manage your Algorand addresses and accounts effortlessly with the AlgoKit Wallet feature. This feature allows you to create short aliases for your addresses and accounts on AlgoKit CLI.

## Usage

Available commands and possible usage as follows:

```bash
$ ~ algokit task wallet
Usage: algokit task wallet [OPTIONS] COMMAND [ARGS]...

Create short aliases for your addresses and accounts on AlgoKit CLI.

Options:
-h, --help Show this message and exit.

Commands:
add Add an address or account to be stored against a named alias.
get Get an address or account stored against a named alias.
list List all addresses and accounts stored against a named alias.
remove Remove an address or account stored against a named alias.
reset Remove all aliases.
```

## Commands

### Add

This command adds an address or account to be stored against a named alias. If the `--mnemonic` flag is used, it will prompt the user for a mnemonic phrase interactively using masked input. If the `--force` flag is used, it will allow overwriting an existing alias. Maximum number of aliases that can be stored at a time is 50.

```bash
$ algokit wallet add [OPTIONS] ALIAS_NAME
```

> Please note, the command is not designed to be used in CI scope, there is no option to skip interactive masked input of the mnemonic, if you want to alias an `Account` (both private and public key) entity.

#### Options

- `--address, -a TEXT`: Specifies the address of the account. This option is required.
- `--mnemonic, -m`: If specified, it prompts the user for a mnemonic phrase interactively using masked input.
- `--force, -f`: If specified, it allows overwriting an existing alias without interactive confirmation prompt.

### Get

This command retrieves an address or account stored against a named alias.

```bash
$ algokit wallet get ALIAS
```

### List

This command lists all addresses and accounts stored against a named alias. If a record contains a `private_key` it will show a boolean flag indicating whether it exists, actual private key values are never exposed. As a user you can obtain the content of the stored aliases by navigating to your dedicated password manager (see [keyring details](https://pypi.org/project/keyring/)).

```bash
$ algokit wallet list
```

### Remove

This command removes an address or account stored against a named alias.
You must confirm the prompt interactively or pass `--force` | `-f` flag to ignore the prompt.

```bash
$ algokit wallet remove ALIAS  [--force | -f]
```

### Reset

This command removes all aliases. You must confirm the prompt interactively or pass `--force` | `-f` flag to ignore the prompt.

```bash
$ algokit wallet reset [--force | -f]
```

## Keyring

AlgoKit relies on the [keyring](https://pypi.org/project/keyring/) library, which provides an easy way to interact with the operating system's password manager. This abstraction allows AlgoKit to securely manage sensitive information such as mnemonics and private keys.

When you use AlgoKit to store a mnemonic, it is never printed or exposed directly in the console. Instead, the mnemonic is converted and stored as a private key in the password manager. This ensures that your sensitive information is kept secure.

To retrieve the stored mnemonic, you will need to manually navigate to your operating system's password manager. The keyring library supports a variety of password managers across different operating systems. Here are some examples:

- On macOS, it uses the Keychain Access app.
- On Windows, it uses the Credential Manager.
- On Linux, it can use Secret Service API, KWallet, or an in-memory store depending on your setup.

> Remember, AlgoKit is designed to keep your sensitive information secure however your storage is only as secure as the device on which it is stored. Always ensure to maintain good security practices on your device, especially when dealing with mnemonics that are to be used on MainNet.

## Further Reading

For in-depth details, visit the [wallet section](../../cli/index.md#wallet) in the AlgoKit CLI reference documentation.
