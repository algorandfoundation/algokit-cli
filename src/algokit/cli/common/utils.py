# Cli/Click related helper functions and classes


import typing as t

import click

from algokit.cli.common.constants import ExplorerEntityType


class MutuallyExclusiveOption(click.Option):
    """
    A Click option that defines mutually exclusive command line options.

    Args:
        *args: Positional arguments passed to the parent class constructor.
        **kwargs: Keyword arguments passed to the parent class constructor.
            not_required_if (list): A list of options that the current option is mutually exclusive with.

    Attributes:
        not_required_if (list): A list of options that the current option is mutually exclusive with.

    Raises:
        AssertionError: If the `not_required_if` parameter is not provided.

    Example:
        ```python
        @click.command()
        @click.option('--option1', help='Option 1')
        @click.option('--option2', help='Option 2')
        @click.option('--option3', help='Option 3', cls=MutuallyExclusiveOption, not_required_if=['option1', 'option2'])
        def my_command(option1, option2, option3):
            # Command logic here
            pass
        ```

        In the example above, the `MutuallyExclusiveOption` class is used to define the `option3` command line option.
        This option is mutually exclusive with `option1` and `option2`,
        meaning that if either `option1` or `option2` is provided, `option3` cannot be used.
        If `option3` is provided along with `option1` or `option2`, a `click.UsageError` is raised.
    """

    def __init__(self, *args: t.Any, **kwargs: t.Any) -> None:
        self.not_required_if: list = kwargs.pop("not_required_if")

        assert self.not_required_if, "'not_required_if' parameter required"
        kwargs["help"] = (
            kwargs.get("help", "") + " Option is mutually exclusive with " + ", ".join(self.not_required_if) + "."
        ).strip()
        super().__init__(*args, **kwargs)

    def handle_parse_result(
        self, ctx: click.Context, opts: t.Mapping[str, t.Any], args: list[str]
    ) -> tuple[t.Any, list[str]]:
        """
        Overrides the `handle_parse_result` method of the `click.Option` class.

        This method checks if the current option is present in the provided options (`opts`)
        and if any of the mutually exclusive options are also present.
        If both the current option and a mutually exclusive option are present, it raises a `click.UsageError`.
        Otherwise, it returns the result of the parent `handle_parse_result` method.

        Args:
            ctx (click.Context): The Click context object.
            opts (dict): The dictionary of parsed options.
            args (list): The list of remaining arguments.

        Returns:
            The result of the parent `handle_parse_result` method.

        Raises:
            click.UsageError: If the current option and a mutually exclusive option are both present.
        """

        current_opt: bool = self.name in opts
        for mutex_opt in self.not_required_if:
            if mutex_opt in opts:
                if current_opt:
                    raise click.UsageError(
                        "Illegal usage: '" + str(self.name) + "' is mutually exclusive with " + str(mutex_opt) + "."
                    )
                self.prompt = None
        return super().handle_parse_result(ctx, opts, args)


def get_explorer_url(identifier: str | int, network: str, entity_type: ExplorerEntityType) -> str:
    """
    Returns a URL for exploring a specified type (transaction, asset, address) on the specified network.

    Args:
        identifier (str | int): The ID of the transaction, asset, or address to explore.
        network (str): The name of the network (e.g., "localnet", "testnet", "mainnet").
        entity_type (ExplorerEntityType): The type to explore (e.g., ExplorerEntityType.TRANSACTION,
        ExplorerEntityType.ASSET, ExplorerEntityType.ADDRESS).

    Returns:
        str: The URL for exploring the specified type on the specified network.

    Raises:
        ValueError: If the network or explorer type is invalid.
    """

    base_urls: dict[str, dict[str, str]] = {
        "testnet": {
            ExplorerEntityType.TRANSACTION.value: "https://testnet.algoexplorer.io/tx/",
            ExplorerEntityType.ASSET.value: "https://testnet.explorer.perawallet.app/assets/",
            ExplorerEntityType.ADDRESS.value: "https://testnet.algoexplorer.io/address/",
        },
        "mainnet": {
            ExplorerEntityType.TRANSACTION.value: "https://algoexplorer.io/tx/",
            ExplorerEntityType.ASSET.value: "https://explorer.perawallet.app/assets/",
            ExplorerEntityType.ADDRESS.value: "https://algoexplorer.io/address/",
        },
    }

    if network == "localnet":
        return f"https://app.dappflow.org/setnetwork?name=sandbox&redirect=explorer/{entity_type.value}/{identifier}/"

    if network not in base_urls:
        raise ValueError(f"Invalid network: {network}")

    if entity_type.value not in base_urls[network]:
        raise ValueError(f"Invalid explorer type: {entity_type}")

    return base_urls[network][entity_type.value] + str(identifier)
