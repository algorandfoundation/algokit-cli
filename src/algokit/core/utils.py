import re
import socket


def extract_version_triple(version_str: str) -> str:
    match = re.search(r"\d+\.\d+\.\d+", version_str)
    if not match:
        raise ValueError("Unable to parse version number")
    return match.group()


def is_minimum_version(system_version: str, minimum_version: str) -> bool:
    system_version_as_tuple = tuple(map(int, system_version.split(".")))
    minimum_version_as_tuple = tuple(map(int, minimum_version.split(".")))
    return system_version_as_tuple >= minimum_version_as_tuple


def is_network_available(host: str = "8.8.8.8", port: int = 53, timeout: float = 3.0) -> bool:
    """
    Check if internet is available by trying to establish a socket connection.
    """

    try:
        socket.setdefaulttimeout(timeout)
        socket.socket(socket.AF_INET, socket.SOCK_STREAM).connect((host, port))
        return True
    except OSError:
        return False
