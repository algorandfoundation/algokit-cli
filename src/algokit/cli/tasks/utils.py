# AlgoKit Tasks related utility functions

import logging
import os
import stat
import sys

import algosdk
import algosdk.encoding
import click
from algokit_utils import (
    Account,
    get_algod_client,
    get_algonode_config,
    get_default_localnet_config,
)
from algosdk.util import algos_to_microalgos, microalgos_to_algos

from algokit.cli.common.constants import AlgorandNetwork
from algokit.core.tasks.wallet import get_alias

logger = logging.getLogger(__name__)


def _validate_asset_balance(account_info: dict, asset_id: int, decimals: int, amount: int = 0) -> None:
    asset_record = next((asset for asset in account_info.get("assets", []) if asset["asset-id"] == asset_id), None)

    if not asset_record:
        raise click.ClickException("Account is not opted into the asset")

    if amount > 0 and asset_record["amount"] < amount:
        required = amount / 10**decimals
        available = asset_record["amount"] / 10**decimals
        raise click.ClickException(
            f"Insufficient asset balance in account, required: {required}, available: {available}"
        )


def _validate_algo_balance(account_info: dict, amount: int) -> None:
    if account_info.get("amount", 0) < amount:
        required = microalgos_to_algos(amount)  # type: ignore[no-untyped-call]
        available = microalgos_to_algos(account_info.get("amount", 0))  # type: ignore[no-untyped-call]
        raise click.ClickException(
            f"Insufficient Algos balance in account, required: {required} Algos, available: {available} Algos"
        )


def get_private_key_from_mnemonic() -> str:
    """
    Converts a mnemonic phrase into a private key.

    Returns:
        str: The private key generated from the mnemonic phrase.

    Raises:
        click.ClickException: If the entered mnemonic phrase is invalid.

    Example Usage:
        private_key = get_private_key_from_mnemonic()
        print(private_key)

    Inputs:
        None

    Flow:
        1. Prompts the user to enter a mnemonic phrase.
        2. Converts the entered mnemonic phrase into a private key using `algosdk.mnemonic.to_private_key`.
        3. Returns the private key.

    """

    mnemonic_phrase = click.prompt("Enter the mnemonic phrase (25 words separated by whitespace)", hide_input=True)
    try:
        return str(algosdk.mnemonic.to_private_key(mnemonic_phrase))  # type: ignore[no-untyped-call]
    except Exception as err:
        raise click.ClickException("Invalid mnemonic. Please provide a valid Algorand mnemonic.") from err


def load_algod_client(network: AlgorandNetwork) -> algosdk.v2client.algod.AlgodClient:
    """
    Returns an instance of the `algosdk.v2client.algod.AlgodClient` class for the specified network.

    Args:
        network (str): The network for which the `AlgodClient` instance needs to be loaded.

    Returns:
        algosdk.v2client.algod.AlgodClient: An instance of the `AlgodClient` class for the specified network.

    Raises:
        click.ClickException: If the specified network is invalid.
    """

    config_mapping = {
        AlgorandNetwork.LOCALNET: get_default_localnet_config("algod"),
        AlgorandNetwork.TESTNET: get_algonode_config("testnet", "algod", ""),
        AlgorandNetwork.MAINNET: get_algonode_config("mainnet", "algod", ""),
    }
    try:
        return get_algod_client(config_mapping[network])
    except KeyError as err:
        raise click.ClickException("Invalid network") from err


def get_asset_decimals(asset_id: int, algod_client: algosdk.v2client.algod.AlgodClient) -> int:
    """
    Retrieves the number of decimal places for a given asset.

    Args:
        asset_id (int): The ID of the asset for which the number of decimal places is to be retrieved. (0 for Algo)
        algod_client (algosdk.v2client.algod.AlgodClient): An instance of the AlgodClient class.

    Returns:
        int: The number of decimal places for the given asset.

    Raises:
        click.ClickException: If the asset info response is invalid.

    Example:
        asset_id = 123
        algod_client = algosdk.v2client.algod.AlgodClient("https://api.algoexplorer.io", "API_KEY")
        decimals = get_asset_decimals(asset_id, algod_client)
        print(decimals)
    """

    if asset_id == 0:
        return 6

    asset_info = algod_client.asset_info(asset_id)

    if not isinstance(asset_info, dict) or "params" not in asset_info or "decimals" not in asset_info["params"]:
        raise click.ClickException("Invalid asset info response")

    return int(asset_info["params"]["decimals"])


def validate_balance(
    algod_client: algosdk.v2client.algod.AlgodClient, account: Account | str, asset_id: int, amount: int = 0
) -> None:
    """
    Validates the balance of an account before an operation.

    Args:
        algod_client (algosdk.v2client.algod.AlgodClient): The AlgodClient object for
        interacting with the Algorand blockchain.
        account (Account | str): The account object.
        asset_id (int): The ID of the asset to be checked (0 for Algos).
        amount (int): The amount of Algos or asset for the operation. Defaults to 0 implying opt-in check only.

    Raises:
        click.ClickException: If any validation check fails.
    """
    address = account.address if isinstance(account, Account) else account
    account_info = algod_client.account_info(address)

    if not isinstance(account_info, dict):
        raise click.ClickException("Invalid account info response")

    if asset_id == 0:
        _validate_algo_balance(account_info, amount)
    else:
        decimals = get_asset_decimals(asset_id, algod_client)
        _validate_asset_balance(account_info, asset_id, decimals, amount)


def validate_address(address: str) -> None:
    """
    Check if a given address is a valid Algorand account address.

    Args:
        address (str): The address to be validated.

    Raises:
        click.ClickException: If the address is invalid.

    Returns:
        None
    """

    if not algosdk.encoding.is_valid_address(address):  # type: ignore[no-untyped-call]
        raise click.ClickException(f"`{address}` is an invalid account address")


def get_account_with_private_key(address: str) -> Account:
    """
    Retrieves an account object with the private key based on the provided address.

    Args:
        address (str): The address for which to retrieve the account object.

    Returns:
        Account: An account object with the address and private key.

    Raises:
        click.ClickException: If the address is not valid or if the alias does not exist or does not have a private key.
    """

    parsed_address = address.strip('"')

    try:
        validate_address(parsed_address)
        pk = get_private_key_from_mnemonic()
        return Account(address=parsed_address, private_key=pk)
    except click.ClickException as ex:
        alias_data = get_alias(parsed_address)

        if not alias_data:
            raise click.ClickException(f"Alias `{parsed_address}` alias does not exist.") from ex
        if not alias_data.private_key:
            raise click.ClickException(f"Alias `{parsed_address}` does not have a private key.") from ex

        return Account(address=alias_data.address, private_key=alias_data.private_key)


def get_address(address: str) -> str:
    """
    Validates the given address and returns it if valid. If the address is not valid,
    it checks if the address is an alias and returns the corresponding address if the alias exists.

    Args:
        address (str): The address to be validated or checked for an alias.

    Returns:
        str: The validated address or the corresponding address from an alias.

    Raises:
        click.ClickException: If the address is not valid and no corresponding alias exists.

    Example:
        address = get_address("ABCD1234")
        print(address)
    """

    parsed_address = address.strip('"')

    try:
        validate_address(parsed_address)
        return parsed_address
    except click.ClickException as ex:
        if len(parsed_address) == algosdk.constants.address_len:
            raise click.ClickException(f"`{parsed_address}` is an invalid account address") from ex

        alias_data = get_alias(parsed_address)

        if not alias_data:
            raise click.ClickException(f"Alias `{parsed_address}` alias does not exist.") from ex

        return alias_data.address


def stdin_has_content() -> bool:
    """
    Checks if there is content in the standard input.

    Returns:
        bool: True if there is content in the standard input, False otherwise.
    """

    mode = os.fstat(sys.stdin.fileno()).st_mode
    return stat.S_ISFIFO(mode) or stat.S_ISREG(mode)


def validate_account_balance_to_opt_in(
    algod_client: algosdk.v2client.algod.AlgodClient, account: Account, num_assets: int
) -> None:
    """
    Validates the balance of an account before opt in operation.
    Each asset requires 0.1 Algos to opt in.

    Args:
        algod_client (algosdk.v2client.algod.AlgodClient): The AlgodClient object for
        interacting with the Algorand blockchain.
        account (Account | str): The account object.
        num_assets (int): The number of the assets for opt in (0 for Algos).

    Raises:
        click.ClickException: If there is an insufficient fund in the account or account is not valid.
    """

    address = account.address if isinstance(account, Account) else account
    account_info = algod_client.account_info(address)

    if not isinstance(account_info, dict):
        raise click.ClickException("Invalid account info response")

    required_microalgos = num_assets * algos_to_microalgos(0.1)  # type: ignore[no-untyped-call]
    available_microalgos = account_info.get("amount", 0)
    if available_microalgos < required_microalgos:
        required_algo = microalgos_to_algos(required_microalgos)  # type: ignore[no-untyped-call]
        available_algos = microalgos_to_algos(available_microalgos)  # type: ignore[no-untyped-call]
        raise click.ClickException(
            f"Insufficient Algos balance in account to opt in, required: {required_algo} Algos, available:"
            f" {available_algos} Algos"
        )


def get_account_info(algod_client: algosdk.v2client.algod.AlgodClient, account_address: str) -> dict:
    account_info = algod_client.account_info(account_address)
    assert isinstance(account_info, dict)
    return account_info
