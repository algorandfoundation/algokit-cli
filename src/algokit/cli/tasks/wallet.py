import json
import re

import click
from algosdk import account

from algokit.cli.tasks.utils import get_private_key_from_mnemonic, validate_address
from algokit.core.tasks.wallet import (
    WALLET_ALIASING_MAX_LIMIT,
    WalletAliasingLimitError,
    add_alias,
    get_alias,
    get_aliases,
    remove_alias,
)


def _validate_alias_name(alias_name: str) -> None:
    pattern = r"^[\w-]{1,20}$"
    if not re.match(pattern, alias_name):
        raise click.ClickException(
            "Invalid alias name. It should have at most 20 characters consisting of numbers, "
            "letters, dashes, or underscores."
        )


@click.group()
def wallet() -> None:
    """Create short aliases for your addresses and accounts on AlgoKit CLI."""


@wallet.command("add")
@click.argument("alias_name", type=click.STRING)
@click.option("--address", "-a", type=click.STRING, required=True, help="The address of the account.")
@click.option(
    "--mnemonic",
    "-m",
    "use_mnemonic",
    is_flag=True,
    help="If specified then prompt the user for a mnemonic phrase interactively using masked input.",
)
@click.option("--force", "-f", is_flag=True, help="Allow overwriting an existing alias.", type=click.BOOL)
def add(*, alias_name: str, address: str, use_mnemonic: bool, force: bool) -> None:
    """Add an address or account to be stored against a named alias (at most 50 aliases)."""

    _validate_alias_name(alias_name)
    validate_address(address)

    private_key = get_private_key_from_mnemonic() if use_mnemonic else None

    if use_mnemonic:
        derived_address = account.address_from_private_key(private_key)  # type: ignore[no-untyped-call]
        if derived_address != address:
            click.echo(
                "Warning: Address from the mnemonic doesn't match the provided address. "
                "It won't work unless the account has been rekeyed."
            )

    if get_alias(alias_name) and not force:
        response = click.prompt(
            f"Alias '{alias_name}' already exists. Overwrite?",
            type=click.Choice(["y", "n"]),
            default="n",
        )
        if response == "n":
            return

    try:
        add_alias(alias_name, address, private_key)
    except WalletAliasingLimitError as ex:
        raise click.ClickException(f"Reached the max of {WALLET_ALIASING_MAX_LIMIT} aliases.") from ex
    except Exception as ex:
        raise click.ClickException("Failed to add alias") from ex
    else:
        click.echo(f"Alias '{alias_name}' added successfully.")


@wallet.command("get")
@click.argument("alias", type=click.STRING)
def get(alias: str) -> None:
    """Get an address or account stored against a named alias."""
    alias_data = get_alias(alias)

    if not alias_data:
        raise click.ClickException(f"Alias `{alias}` does not exist.")

    click.echo(
        f"Address for alias `{alias}`: {alias_data.address}"
        f"{' (🔐 includes private key)' if alias_data.private_key else ''}"
    )


@wallet.command("list")
def list_all() -> None:
    """List all addresses and accounts stored against a named alias."""

    aliases = get_aliases()

    output = [
        {
            "alias": alias_data.alias,
            "address": alias_data.address,
            "has_private_key": bool(alias_data.private_key),
        }
        for alias_data in aliases
    ]

    content = (
        json.dumps(output, indent=2)
        if output
        else "You don't have any aliases stored yet. Create one using `algokit task wallet add`."
    )

    click.echo(content)


@wallet.command("remove")
@click.argument("alias", type=click.STRING)
@click.option("--force", "-f", is_flag=True, help="Allow removing an alias without confirmation.")
def remove(*, alias: str, force: bool) -> None:
    """Remove an address or account stored against a named alias."""

    alias_data = get_alias(alias)

    if not alias_data:
        raise click.ClickException(f"Alias `{alias}` does not exist.")

    if not force:
        response = click.prompt(
            f"🚨 This is a destructive action that will remove the `{alias_data.alias}` alias. Are you sure?",
            type=click.Choice(["y", "n"]),
            default="n",
        )

        if response == "n":
            return

    remove_alias(alias)

    click.echo(f"Alias `{alias}` removed successfully.")


@wallet.command("reset")
@click.option("--force", "-f", is_flag=True, help="Allow removing all aliases without confirmation.")
def reset(*, force: bool) -> None:
    """Remove all aliases."""

    aliases = get_aliases()

    if not aliases:
        click.echo("Warning: No aliases available to reset.")
        return

    if not force:
        response = click.prompt(
            "🚨 This is a destructive action that will clear all aliases. Are you sure?",
            type=click.Choice(["y", "n"]),
            default="n",
        )

        if response == "n":
            return

    for alias_data in aliases:
        try:
            remove_alias(alias_data.alias)
        except Exception as ex:
            raise click.ClickException(f"Failed to remove alias {alias_data.alias}") from ex

    click.echo("All aliases have been cleared.")
