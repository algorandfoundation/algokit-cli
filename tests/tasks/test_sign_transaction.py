import json

import click
import pytest
from algosdk import encoding, mnemonic, transaction

from algokit.core.tasks.wallet import WALLET_ALIASES_KEYRING_USERNAME
from tests.tasks.conftest import DUMMY_ACCOUNT, DUMMY_SUGGESTED_PARAMS
from tests.utils.approvals import verify
from tests.utils.click_invoker import invoke


def _generate_dummy_txn(sender: str, amount: int = 1) -> transaction.PaymentTxn:
    return transaction.PaymentTxn(sender, DUMMY_SUGGESTED_PARAMS, sender, amt=amount)  # type: ignore[no-untyped-call]


def _get_mnemonic_from_private_key(private_key: str) -> str:
    return str(mnemonic.from_private_key(private_key))  # type: ignore[no-untyped-call]


def test_sign_atomic_txn_group_successful(
    tmp_path_factory: pytest.TempPathFactory, mock_keyring: dict[str, str]
) -> None:
    # Arrange
    cwd = tmp_path_factory.mktemp("cwd")
    alias_name = "dummy_alias"
    mock_keyring[alias_name] = json.dumps(
        {
            "alias": alias_name,
            "address": DUMMY_ACCOUNT.address,
            "private_key": DUMMY_ACCOUNT.private_key,
        }
    )
    mock_keyring[WALLET_ALIASES_KEYRING_USERNAME] = json.dumps([alias_name])
    txn_a = _generate_dummy_txn(DUMMY_ACCOUNT.address)
    txn_b = _generate_dummy_txn(DUMMY_ACCOUNT.address)
    gid = transaction.calculate_group_id([txn_a, txn_b])  # type: ignore[no-untyped-call]
    txn_a.group = gid
    txn_b.group = gid
    transaction.write_to_file([txn_a, txn_b], str(cwd / "dummy.txns"))  # type: ignore[no-untyped-call]

    # Act
    result = invoke(f"task sign -a {alias_name} --file dummy.txns", input="y", cwd=cwd)

    # Assert
    assert result.exit_code == 0
    verify(result.output)


def test_sign_from_stdin_with_alias_successful(mock_keyring: dict[str, str]) -> None:
    # Arrange
    alias_name = "dummy_alias"
    mock_keyring[alias_name] = json.dumps(
        {
            "alias": alias_name,
            "address": DUMMY_ACCOUNT.address,
            "private_key": DUMMY_ACCOUNT.private_key,
        }
    )
    mock_keyring[WALLET_ALIASES_KEYRING_USERNAME] = json.dumps([alias_name])
    dummy_txn = _generate_dummy_txn(DUMMY_ACCOUNT.address)

    # Act
    txn = encoding.msgpack_encode({"txn": dummy_txn.dictify()})  # type: ignore[no-untyped-call]
    result = invoke(f"task sign -a {alias_name} --transaction {txn}", input="y")

    # Assert
    assert result.exit_code == 0
    verify(result.output)


def test_sign_from_stdin_with_address_successful() -> None:
    # Arrange
    dummy_txn = _generate_dummy_txn(DUMMY_ACCOUNT.address)

    # Act
    txn = encoding.msgpack_encode({"txn": dummy_txn.dictify()})  # type: ignore[no-untyped-call]
    result = invoke(
        f"task sign -a {DUMMY_ACCOUNT.address} --transaction {txn}",
        input=f"{_get_mnemonic_from_private_key(DUMMY_ACCOUNT.private_key)}\ny",
    )

    # Assert
    assert result.exit_code == 0
    verify(result.output)


def test_sign_many_from_file_with_alias_successful(
    tmp_path_factory: pytest.TempPathFactory, mock_keyring: dict[str, str]
) -> None:
    # Arrange
    cwd = tmp_path_factory.mktemp("cwd")
    alias_name = "dummy_alias"
    mock_keyring[alias_name] = json.dumps(
        {
            "alias": alias_name,
            "address": DUMMY_ACCOUNT.address,
            "private_key": DUMMY_ACCOUNT.private_key,
        }
    )
    mock_keyring[WALLET_ALIASES_KEYRING_USERNAME] = json.dumps([alias_name])
    _generate_dummy_txn(DUMMY_ACCOUNT.address)
    transaction.write_to_file(  # type: ignore[no-untyped-call]
        [
            _generate_dummy_txn(DUMMY_ACCOUNT.address, 1),
            _generate_dummy_txn(DUMMY_ACCOUNT.address, 2),
            _generate_dummy_txn(DUMMY_ACCOUNT.address, 3),
        ],
        str(cwd / "dummy.txns"),
    )

    # Act
    result = invoke(f"task sign -a {alias_name} --file dummy.txns", input="y", cwd=cwd)

    # Assert
    assert result.exit_code == 0
    verify(result.output)


def test_sign_many_from_file_with_address_successful(
    tmp_path_factory: pytest.TempPathFactory,
) -> None:
    # Arrange
    cwd = tmp_path_factory.mktemp("cwd")
    _generate_dummy_txn(DUMMY_ACCOUNT.address)
    transaction.write_to_file(  # type: ignore[no-untyped-call]
        [
            _generate_dummy_txn(DUMMY_ACCOUNT.address, 1),
            _generate_dummy_txn(DUMMY_ACCOUNT.address, 2),
            _generate_dummy_txn(DUMMY_ACCOUNT.address, 3),
        ],
        str(cwd / "dummy.txns"),
    )

    # Act
    result = invoke(
        f"task sign -a {DUMMY_ACCOUNT.address} --file dummy.txns",
        input=f"{_get_mnemonic_from_private_key(DUMMY_ACCOUNT.private_key)}\ny",
        cwd=cwd,
    )

    # Assert
    assert result.exit_code == 0
    verify(result.output)


def test_mutually_exclusive_options() -> None:
    # Arrange
    _generate_dummy_txn(DUMMY_ACCOUNT.address)

    # Act
    result = invoke(
        f"task sign -a {DUMMY_ACCOUNT.address} --file dummy.txns --transaction dummy.txn",
        input=f"{_get_mnemonic_from_private_key(DUMMY_ACCOUNT.private_key)}\ny",
    )

    # Assert
    assert result.exit_code == click.exceptions.UsageError.exit_code
    verify(result.output)


def test_file_decoding_errors(tmp_path_factory: pytest.TempPathFactory) -> None:
    # Arrange
    cwd = tmp_path_factory.mktemp("cwd")
    (cwd / "dummy.txns").touch()

    # Act
    result = invoke(
        f"task sign -a {DUMMY_ACCOUNT.address} --file dummy.txns",
        input=f"{_get_mnemonic_from_private_key(DUMMY_ACCOUNT.private_key)}\ny",
        cwd=cwd,
    )

    # Assert
    assert result.exit_code == 1
    verify(result.output)


def test_transaction_decoding_errors() -> None:
    # Act
    result = invoke(
        f"task sign -a {DUMMY_ACCOUNT.address} --transaction dummy",
        input=f"{_get_mnemonic_from_private_key(DUMMY_ACCOUNT.private_key)}\ny",
    )

    # Assert
    assert result.exit_code == 1
    verify(result.output)
