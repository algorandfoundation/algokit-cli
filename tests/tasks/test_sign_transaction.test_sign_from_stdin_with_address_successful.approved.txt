Enter the mnemonic phrase (25 words separated by whitespace): 
[
  {
    "transaction_id": "YGSOMX5QSASQALR5V4MH47L4GEHFJZVGSUETJAWUFM2I6PJAQL4Q",
    "content": {
      "amt": 1,
      "fv": 33652328,
      "gen": "testnet-v1.0",
      "gh": "SGO1GKSzyE7IEPItTxCByw9x8FmnrCDexi9/cOUJOiI=",
      "lv": 33653328,
      "rcv": "ZbpO9Kb7j1+HWA9SYq/ObmT/gCJjP/EgZnT0aQUkxQI=",
      "snd": "ZbpO9Kb7j1+HWA9SYq/ObmT/gCJjP/EgZnT0aQUkxQI=",
      "type": "pay"
    }
  }
]
Would you like to proceed with signing the above? (y, n) [n]: y
[
  {
    "transaction_id": "YGSOMX5QSASQALR5V4MH47L4GEHFJZVGSUETJAWUFM2I6PJAQL4Q",
    "content": "gqNzaWfEQHHuPbhkAADyq8eKU/NiDCuJ+cnW9MrHT3iAAEPm+okmoio/rmgctA7QUpqqd4eF5aYlUcgz9EbU+uUXS1rFOg2jdHhuiKNhbXQBomZ2zgIBfmijZ2VurHRlc3RuZXQtdjEuMKJnaMQgSGO1GKSzyE7IEPItTxCByw9x8FmnrCDexi9/cOUJOiKibHbOAgGCUKNyY3bEIGW6TvSm+49fh1gPUmKvzm5k/4AiYz/xIGZ09GkFJMUCo3NuZMQgZbpO9Kb7j1+HWA9SYq/ObmT/gCJjP/EgZnT0aQUkxQKkdHlwZaNwYXk="
  }
]
